src/action.ts

```diff
@@ -116,57 +143,180 @@ async function setupBinaryInEnv(binary: DownloadedBinary) {
+        );
+      } else {
+        throw error;
+      }
+    }
+  }
+}
 
-  const getReleaseResponse = await octokit.rest.repos.getReleaseByTag({
-    owner,
-    repo,
-    tag,
-  });
+function createGitHubProvider(config: GitHubConfig): GitHubProviderInterface {
+  return new GitHubProvider(config);
+}
 
+function findCompatibleAsset(release: Release, owner: string, repo: string, tag: string): ReleaseAsset {
   const re = new RegExp(`${osPlatform()}.*${arch()}`);
-  const asset = getReleaseResponse.data.assets.find((obj: any) => re.test(obj.name));
+  const asset = release.assets.find((obj: ReleaseAsset) => re.test(obj.name));
 
   if (!asset) {
     throw new Error(`Can not find ${owner}/${repo} release for ${tag} in platform ${re}.`);
   }
 
-  // Use @actions/tool-cache to download the binary from GitHub
-  const downloadedPath = await toolCache.downloadTool(
-    asset.url,
-    // Don't set a destination path. It will default to a temporary one.
-    undefined,
-    `token ${token}`,
-    {
-      accept: "application/octet-stream",
-    }
-  );
+  return asset;
+}
 
-  core.debug(`${owner}/${repo}@'${tag}' has been downloaded at ${downloadedPath}`);
+function determineDownloadConfig(
+  asset: ReleaseAsset,
+  repo: string,
+  token: string
+): { assetUrl: string; authHeader: string | undefined; headers: Record<string, string> } {
+  // Always use the API URL for consistent authentication behavior
+  const assetUrl = asset.url;
+
+  core.debug(`Selected asset URL for ${assetUrl} (using API URL)`);
+
+  // Determine authentication and headers
+  let authHeader: string | undefined;
+  const headers: Record<string, string> = {};
+
+  // Asset API URL: always needs proper authentication
+  if (token) {
+    authHeader = `Bearer ${token}`;
+  }
+  // Asset API requires specific headers
+  headers.accept = "application/octet-stream";
+  headers["X-GitHub-Api-Version"] = "2022-11-28";
+
+  return { assetUrl, authHeader, headers };
+}
+
+async function downloadAssetWithRetry(
+  assetUrl: string,
+  authHeader: string | undefined,
+  headers: Record<string, string>,
+  owner: string,
+  repo: string,
+  tag: string,
+  token: string
+): Promise<string> {
+  const isPublicTool = PUBLIC_TOOLS.includes(repo as any);
+
+  // For public tools, temporarily unset GITHUB_BASE_URL to avoid protocol conflicts
+  const originalGithubBaseUrl = process.env.GITHUB_BASE_URL;
+  if (isPublicTool) {
+    delete process.env.GITHUB_BASE_URL;
+  }
+
+  try {
+    return await toolCache.downloadTool(assetUrl, undefined, authHeader, headers);
+  } catch (err: any) {
+    const status = (err?.status || err?.code || "").toString();
+    const isAuthIssue = status === "404" || status === "403";
+
+    if (isPublicTool && authHeader && isAuthIssue) {
```

Comment:
NIT: This block is kind of hard to read. Could we replace the else's with early returns/throws?